CSharp session caching functionality

nikoo28 opened the pull request:
- Using Microsoft `InMemoryCache` to cache sessions
  - Using time and size based eviction policies.
  - Since the sessions are stateful, we only remove them from the cache when they are no longer used.
  - Added comments to explain usage.
  - the actual cached object is `EnvelopeEncryptionJsonImpl` due to generics/simplicity, but we're effectively caching the session objects (and thereby the IKs)

- Added 3 new options to CryptoPolicy
  - _canCacheSessions_: defaults to false to maintain backwards compatibility
  - _sessionCacheMaxSize_
  - _sessionCacheExpireMillis_
  - added documentation around BasicExpiringCryptoPolicy builder to highlight the current defaults

- Added unit tests for eviction scenarios and tests to verify the cache size at desired times
- removed `SecureCryptoKeyDictionaryFactory` (wasn't really needed, especially with these changes)
- some minor refactor and cleanup.

maximgladkov commented:
# [Codecov](https://codecov.io/gh/godaddy/asherah/pull/118?src=pr&el=h1) Report
> Merging [#118](https://codecov.io/gh/godaddy/asherah/pull/118?src=pr&el=desc) into [master](https://codecov.io/gh/godaddy/asherah/commit/7db1c6dd4fc000c494497994fc60bc102b59968b&el=desc) will **decrease** coverage by `0.57%`.
> The diff coverage is `n/a`.

[![Impacted file tree graph](https://codecov.io/gh/godaddy/asherah/pull/118/graphs/tree.svg?width=650&height=150&src=pr&token=SJM3rBBurP)](https://codecov.io/gh/godaddy/asherah/pull/118?src=pr&el=tree)

```diff
@@             Coverage Diff              @@
##             master     #118      +/-   ##
============================================
- Coverage     97.05%   96.47%   -0.58%     
  Complexity      399      399              
============================================
  Files           105       66      -39     
  Lines          2476     1391    -1085     
  Branches         60       60              
============================================
- Hits           2403     1342    -1061     
+ Misses           63       39      -24     
  Partials         10       10              
```

| Flag | Coverage Δ | Complexity Δ | |
|---|---|---|---|
| #languages_csharp_AppEncryption | `?` | `?` | |
| #languages_csharp_Logging | `100.00% <ø> (ø)` | `0.00 <ø> (ø)` | :arrow_up: |
| #languages_csharp_SecureMemory | `95.51% <ø> (ø)` | `0.00 <ø> (ø)` | :arrow_up: |
| #languages_java_app_encryption | `96.93% <ø> (ø)` | `330.00 <ø> (ø)` | :arrow_up: |
| #languages_java_secure_memory | `94.91% <ø> (ø)` | `69.00 <ø> (ø)` | :arrow_up: |

| [Impacted Files](https://codecov.io/gh/godaddy/asherah/pull/118?src=pr&el=tree) | Coverage Δ | Complexity Δ | |
|---|---|---|---|
| [...ncryption/AppEncryption/Persistence/Persistence.cs](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree#diff-bGFuZ3VhZ2VzL2NzaGFycC9BcHBFbmNyeXB0aW9uL0FwcEVuY3J5cHRpb24vUGVyc2lzdGVuY2UvUGVyc2lzdGVuY2UuY3M=) | | | |
| [...rp/AppEncryption/AppEncryption/SessionBytesImpl.cs](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree#diff-bGFuZ3VhZ2VzL2NzaGFycC9BcHBFbmNyeXB0aW9uL0FwcEVuY3J5cHRpb24vU2Vzc2lvbkJ5dGVzSW1wbC5jcw==) | | | |
| [...n/Crypto/Exceptions/CipherNotSupportedException.cs](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree#diff-bGFuZ3VhZ2VzL2NzaGFycC9BcHBFbmNyeXB0aW9uL0NyeXB0by9FeGNlcHRpb25zL0NpcGhlck5vdFN1cHBvcnRlZEV4Y2VwdGlvbi5jcw==) | | | |
| [...ppEncryption/Crypto/Envelope/AeadEnvelopeCrypto.cs](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree#diff-bGFuZ3VhZ2VzL2NzaGFycC9BcHBFbmNyeXB0aW9uL0NyeXB0by9FbnZlbG9wZS9BZWFkRW52ZWxvcGVDcnlwdG8uY3M=) | | | |
| [...yption/Crypto/Exceptions/AppEncryptionException.cs](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree#diff-bGFuZ3VhZ2VzL2NzaGFycC9BcHBFbmNyeXB0aW9uL0NyeXB0by9FeGNlcHRpb25zL0FwcEVuY3J5cHRpb25FeGNlcHRpb24uY3M=) | | | |
| [...ages/csharp/AppEncryption/Crypto/NonceGenerator.cs](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree#diff-bGFuZ3VhZ2VzL2NzaGFycC9BcHBFbmNyeXB0aW9uL0NyeXB0by9Ob25jZUdlbmVyYXRvci5jcw==) | | | |
| [...sharp/AppEncryption/Crypto/Keys/SecretCryptoKey.cs](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree#diff-bGFuZ3VhZ2VzL2NzaGFycC9BcHBFbmNyeXB0aW9uL0NyeXB0by9LZXlzL1NlY3JldENyeXB0b0tleS5jcw==) | | | |
| [...ages/csharp/AppEncryption/AppEncryption/Session.cs](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree#diff-bGFuZ3VhZ2VzL2NzaGFycC9BcHBFbmNyeXB0aW9uL0FwcEVuY3J5cHRpb24vU2Vzc2lvbi5jcw==) | | | |
| [...ncryption/AppEncryption/Kms/AwsKmsClientFactory.cs](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree#diff-bGFuZ3VhZ2VzL2NzaGFycC9BcHBFbmNyeXB0aW9uL0FwcEVuY3J5cHRpb24vS21zL0F3c0ttc0NsaWVudEZhY3RvcnkuY3M=) | | | |
| [...es/csharp/AppEncryption/AppEncryption/Partition.cs](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree#diff-bGFuZ3VhZ2VzL2NzaGFycC9BcHBFbmNyeXB0aW9uL0FwcEVuY3J5cHRpb24vUGFydGl0aW9uLmNz) | | | |
| ... and [20 more](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree-more) | |

------

[Continue to review full report at Codecov](https://codecov.io/gh/godaddy/asherah/pull/118?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/godaddy/asherah/pull/118?src=pr&el=footer). Last update [7db1c6d...6c29a56](https://codecov.io/gh/godaddy/asherah/pull/118?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

sushantmimani commented on languages/csharp/AppEncryption/AppEncryption.Tests/AppEncryption/SessionFactoryTest.cs:
```diff
@@ -56,13 +56,682 @@ private void TestConstructor()
+                });
+
+                // Wait for all threads to complete
+                Assert.Equal(numRequests, completedTasks);
```
Reckon that, like Java, we cannot reliably assert the cache size here? If so, might be worth adding a comment about the same. If not, verify cache size here.

sushantmimani commented on languages/csharp/AppEncryption/AppEncryption.Tests/AppEncryption/SessionFactoryTest.cs:
```diff
@@ -56,13 +56,682 @@ private void TestConstructor()
+                        Assert.Equal(payload, decryptedPayload);
+                        Interlocked.Increment(ref completedTasks);
+
+                        Thread.Sleep((int)sessionCacheExpireMillis * 3);
```
Add comment about sleep

sushantmimani commented on languages/csharp/AppEncryption/AppEncryption.Tests/AppEncryption/SessionFactoryTest.cs:
```diff
@@ -56,13 +56,682 @@ private void TestConstructor()
+
+                        Assert.Equal(payload, decryptedPayload);
+                        Interlocked.Increment(ref completedTasks);
+
```
ditto

nikoo28 commented on languages/csharp/AppEncryption/AppEncryption.Tests/AppEncryption/SessionFactoryTest.cs:
```diff
@@ -56,13 +56,682 @@ private void TestConstructor()
+                    }
+
+                    // Even after timeout, verify that we have one entry in cache
+                    Assert.Equal(1, factory.SessionCache.Count);
```
`SessionCache.Count` does an operation on cache.

nikoo28 commented on languages/csharp/AppEncryption/AppEncryption/SessionFactory.cs:
```diff
@@ -8,38 +12,44 @@
     public class SessionFactory : IDisposable
     {
+        #pragma warning disable SA1401
+        protected internal readonly MemoryCache SessionCache;
```
we don't want MemoryCache to be public

nikoo28 commented on languages/csharp/AppEncryption/AppEncryption/SessionFactory.cs:
```diff
@@ -128,18 +157,173 @@ internal Partition GetPartition(string partitionId)
+                    cryptoPolicy,
+                    keyManagementService);
+
+                return new CachedSession(envelopeEncryptionJsonImpl, partitionId, this);
```
this was because we needed the `partitionId` and an instance of `this` in the inner class. discussed briefly about this offline, can clarify further if required.

sushantmimani replied:
Code changed to be consistent with Java

nikoo28 commented on languages/csharp/AppEncryption/AppEncryption.Tests/AppEncryption/SessionFactoryTest.cs:
```diff
@@ -56,13 +56,682 @@ private void TestConstructor()
+
+                    // we should not hit the metastore since the session should not have been evicted
+                    metastoreSpy.Verify(
+                        x => x.Load(It.IsAny<string>(), It.IsAny<DateTimeOffset>()), Times.Never);
```
the metastore was not hit, it inherently verifies that the item was present in the cache.

nikoo28 commented on languages/csharp/AppEncryption/AppEncryption.Tests/AppEncryption/SessionFactoryTest.cs:
```diff
@@ -56,13 +56,682 @@ private void TestConstructor()
+                    // we should hit the metastore since the session should have been evicted as the size limit was reached
+                    metastoreSpy.Verify(
+                        x => x.Load(It.IsAny<string>(), It.IsAny<DateTimeOffset>()));
+                }
```
the test scenario tests if the original entry was evicted from cache. So we are more concerned if the actual session was evicted (hence the metastore hit).

nikoo28 commented on languages/csharp/AppEncryption/AppEncryption.Tests/AppEncryption/SessionFactoryTest.cs:
```diff
@@ -56,13 +56,682 @@ private void TestConstructor()
+                        byte[] decryptedPayload = sessionBytes.Decrypt(drr1);
+
+                        Assert.Equal(payload, decryptedPayload);
+
```
we are validating if the metastore was not hit. it inherently means that the session was present in the cache.

nikoo28 commented:
Addressed review comments.
The remaining comment is around verifying the number of elements in the cache in unit tests. These are being inherently verified with the metastore hits. If we feel these counts are necessary, the test changes can be made in both java and c# at a later stage.

maximgladkov commented:
# [Codecov](https://codecov.io/gh/godaddy/asherah/pull/118?src=pr&el=h1) Report
> Merging [#118](https://codecov.io/gh/godaddy/asherah/pull/118?src=pr&el=desc) into [master](https://codecov.io/gh/godaddy/asherah/commit/7db1c6dd4fc000c494497994fc60bc102b59968b&el=desc) will **decrease** coverage by `0.06%`.
> The diff coverage is `n/a`.

[![Impacted file tree graph](https://codecov.io/gh/godaddy/asherah/pull/118/graphs/tree.svg?width=650&height=150&src=pr&token=SJM3rBBurP)](https://codecov.io/gh/godaddy/asherah/pull/118?src=pr&el=tree)

```diff
@@             Coverage Diff              @@
##             master     #118      +/-   ##
============================================
- Coverage     97.05%   96.98%   -0.07%     
  Complexity      399      399              
============================================
  Files           105      104       -1     
  Lines          2476     2552      +76     
  Branches         60       60              
============================================
+ Hits           2403     2475      +72     
- Misses           63       67       +4     
  Partials         10       10              
```

| Flag | Coverage Δ | Complexity Δ | |
|---|---|---|---|
| #languages_csharp_AppEncryption | `97.58% <94.17%> (-0.20%)` | `0.00 <0.00> (ø)` | |
| #languages_csharp_Logging | `100.00% <ø> (ø)` | `0.00 <ø> (ø)` | :arrow_up: |
| #languages_csharp_SecureMemory | `95.51% <ø> (ø)` | `0.00 <ø> (ø)` | :arrow_up: |
| #languages_java_app_encryption | `96.93% <ø> (ø)` | `330.00 <ø> (ø)` | :arrow_up: |
| #languages_java_secure_memory | `94.91% <ø> (ø)` | `69.00 <ø> (ø)` | :arrow_up: |

| [Impacted Files](https://codecov.io/gh/godaddy/asherah/pull/118?src=pr&el=tree) | Coverage Δ | Complexity Δ | |
|---|---|---|---|
| [...ppEncryption/Kms/StaticKeyManagementServiceImpl.cs](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree#diff-bGFuZ3VhZ2VzL2NzaGFycC9BcHBFbmNyeXB0aW9uL0FwcEVuY3J5cHRpb24vS21zL1N0YXRpY0tleU1hbmFnZW1lbnRTZXJ2aWNlSW1wbC5jcw==) | `100.00% <0.00%> (+25.00%)` | `0.00% <0.00%> (ø%)` | :arrow_up: |

------

[Continue to review full report at Codecov](https://codecov.io/gh/godaddy/asherah/pull/118?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/godaddy/asherah/pull/118?src=pr&el=footer). Last update [7db1c6d...776a3ae](https://codecov.io/gh/godaddy/asherah/pull/118?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

aka-bo commented on languages/csharp/AppEncryption/AppEncryption/SessionFactory.cs:
```diff
@@ -128,18 +160,172 @@ internal Partition GetPartition(string partitionId)
+        private CachedSession AcquireShared(
+            Func<EnvelopeEncryptionJsonImpl> createSessionFunc, string partitionId)
+        {
+            SemaphoreSlim getCachedItemLock = semaphoreLocks.GetOrAdd(partitionId, k => new SemaphoreSlim(1, 1));
```
Seems to me that a semaphore with an initial and maximum requests set to 1 would be functionally equivalent to `lock(getCachedItemLock)`. Is there a benefit to using a `SemaphoreSlim(1, 1)` here that I might be missing? If not, replacing it with a simple `Object()` with `lock` will simplify the code a bit and likely improve performance.

nikoo28 replied:
i remembered that semaphoreslim provided better performance than simple lock, but i think in case of setting the limit to `1`, using just a `lock` would be a better option. Would make these changes soon.

maximgladkov commented:
# [Codecov](https://codecov.io/gh/godaddy/asherah/pull/118?src=pr&el=h1) Report
> Merging [#118](https://codecov.io/gh/godaddy/asherah/pull/118?src=pr&el=desc) into [master](https://codecov.io/gh/godaddy/asherah/commit/7db1c6dd4fc000c494497994fc60bc102b59968b&el=desc) will **decrease** coverage by `0.04%`.
> The diff coverage is `n/a`.

[![Impacted file tree graph](https://codecov.io/gh/godaddy/asherah/pull/118/graphs/tree.svg?width=650&height=150&src=pr&token=SJM3rBBurP)](https://codecov.io/gh/godaddy/asherah/pull/118?src=pr&el=tree)

```diff
@@             Coverage Diff              @@
##             master     #118      +/-   ##
============================================
- Coverage     97.05%   97.00%   -0.05%     
+ Complexity      399       69     -330     
============================================
  Files           105       66      -39     
  Lines          2476     1570     -906     
  Branches         60        8      -52     
============================================
- Hits           2403     1523     -880     
+ Misses           63       42      -21     
+ Partials         10        5       -5     
```

| Flag | Coverage Δ | Complexity Δ | |
|---|---|---|---|
| #languages_csharp_AppEncryption | `97.58% <94.11%> (-0.21%)` | `0.00 <0.00> (ø)` | |
| #languages_csharp_Logging | `100.00% <ø> (ø)` | `0.00 <ø> (ø)` | :arrow_up: |
| #languages_csharp_SecureMemory | `95.51% <ø> (ø)` | `0.00 <ø> (ø)` | :arrow_up: |
| #languages_java_app_encryption | `?` | `?` | |
| #languages_java_secure_memory | `94.91% <ø> (ø)` | `69.00 <ø> (ø)` | :arrow_up: |

| [Impacted Files](https://codecov.io/gh/godaddy/asherah/pull/118?src=pr&el=tree) | Coverage Δ | Complexity Δ | |
|---|---|---|---|
| [.../java/com/godaddy/asherah/crypto/CryptoPolicy.java](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree#diff-bGFuZ3VhZ2VzL2phdmEvYXBwLWVuY3J5cHRpb24vc3JjL21haW4vamF2YS9jb20vZ29kYWRkeS9hc2hlcmFoL2NyeXB0by9DcnlwdG9Qb2xpY3kuamF2YQ==) | | | |
| [...in/java/com/godaddy/asherah/crypto/AeadCrypto.java](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree#diff-bGFuZ3VhZ2VzL2phdmEvYXBwLWVuY3J5cHRpb24vc3JjL21haW4vamF2YS9jb20vZ29kYWRkeS9hc2hlcmFoL2NyeXB0by9BZWFkQ3J5cHRvLmphdmE=) | | | |
| [.../godaddy/asherah/appencryption/SessionFactory.java](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree#diff-bGFuZ3VhZ2VzL2phdmEvYXBwLWVuY3J5cHRpb24vc3JjL21haW4vamF2YS9jb20vZ29kYWRkeS9hc2hlcmFoL2FwcGVuY3J5cHRpb24vU2Vzc2lvbkZhY3RvcnkuamF2YQ==) | | | |
| [...asherah/crypto/bufferutils/ManagedBufferUtils.java](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree#diff-bGFuZ3VhZ2VzL2phdmEvYXBwLWVuY3J5cHRpb24vc3JjL21haW4vamF2YS9jb20vZ29kYWRkeS9hc2hlcmFoL2NyeXB0by9idWZmZXJ1dGlscy9NYW5hZ2VkQnVmZmVyVXRpbHMuamF2YQ==) | | | |
| [...asherah/appencryption/exceptions/KmsException.java](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree#diff-bGFuZ3VhZ2VzL2phdmEvYXBwLWVuY3J5cHRpb24vc3JjL21haW4vamF2YS9jb20vZ29kYWRkeS9hc2hlcmFoL2FwcGVuY3J5cHRpb24vZXhjZXB0aW9ucy9LbXNFeGNlcHRpb24uamF2YQ==) | | | |
| [...odaddy/asherah/crypto/keys/SecureCryptoKeyMap.java](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree#diff-bGFuZ3VhZ2VzL2phdmEvYXBwLWVuY3J5cHRpb24vc3JjL21haW4vamF2YS9jb20vZ29kYWRkeS9hc2hlcmFoL2NyeXB0by9rZXlzL1NlY3VyZUNyeXB0b0tleU1hcC5qYXZh) | | | |
| [...addy/asherah/crypto/BasicExpiringCryptoPolicy.java](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree#diff-bGFuZ3VhZ2VzL2phdmEvYXBwLWVuY3J5cHRpb24vc3JjL21haW4vamF2YS9jb20vZ29kYWRkeS9hc2hlcmFoL2NyeXB0by9CYXNpY0V4cGlyaW5nQ3J5cHRvUG9saWN5LmphdmE=) | | | |
| [...asherah/appencryption/kms/AwsKmsClientFactory.java](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree#diff-bGFuZ3VhZ2VzL2phdmEvYXBwLWVuY3J5cHRpb24vc3JjL21haW4vamF2YS9jb20vZ29kYWRkeS9hc2hlcmFoL2FwcGVuY3J5cHRpb24va21zL0F3c0ttc0NsaWVudEZhY3RvcnkuamF2YQ==) | | | |
| [...m/godaddy/asherah/crypto/keys/SecretCryptoKey.java](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree#diff-bGFuZ3VhZ2VzL2phdmEvYXBwLWVuY3J5cHRpb24vc3JjL21haW4vamF2YS9jb20vZ29kYWRkeS9hc2hlcmFoL2NyeXB0by9rZXlzL1NlY3JldENyeXB0b0tleS5qYXZh) | | | |
| [...ava/com/godaddy/asherah/appencryption/Session.java](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree#diff-bGFuZ3VhZ2VzL2phdmEvYXBwLWVuY3J5cHRpb24vc3JjL21haW4vamF2YS9jb20vZ29kYWRkeS9hc2hlcmFoL2FwcGVuY3J5cHRpb24vU2Vzc2lvbi5qYXZh) | | | |
| ... and [29 more](https://codecov.io/gh/godaddy/asherah/pull/118/diff?src=pr&el=tree-more) | |

------

[Continue to review full report at Codecov](https://codecov.io/gh/godaddy/asherah/pull/118?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/godaddy/asherah/pull/118?src=pr&el=footer). Last update [7db1c6d...d3ca242](https://codecov.io/gh/godaddy/asherah/pull/118?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).
